flang/lib/Lower/IO.cpp

```diff
@@ -324,28 +324,28 @@ static void genInputItemList(Fortran::lower::AbstractConverter &converter,
     auto loc = converter.genLocation(pVar.GetSource());
     makeNextConditionalOn(builder, loc, insertPt, checkResult, ok,
                           inIterWhileLoop);
-    auto itemBox =
+    Fortran::lower::CharacterExprHelper charHelper{builder, loc};
+    auto unsafeItemBox =
         converter.genExprAddr(Fortran::semantics::GetExpr(pVar), loc);
+    auto itemBox = charHelper.cleanUpCharacterExtendedValue(unsafeItemBox);
     auto itemAddr = fir::getBase(itemBox);
-    if (auto box = dyn_cast<fir::EmboxCharOp>(itemAddr.getDefiningOp()))
-      itemAddr = box.memref();
     auto itemTy = fir::dyn_cast_ptrEleTy(itemAddr.getType());
-    if (!itemTy)
+    if (!itemTy) {
       mlir::emitError(loc, "internal: unhandled input item type ")
           << itemAddr.getType();
+      return;
```

Comment:
Is this `return` really required? After the call to `emitError`, we know the IR being generated is broken already.

`emitError` is different in terms of semantics than, say, `assert`. While `assert` is going to crash and burn (for simplicity), `emitError` is meant to continue and queue up a list of errors. When lowering completes, the entire stack of errors is presented to the user, so it can catch more than just the first one.

Reply:
Yes, otherwise it just prints the error and immediately crashes two lines below when using itemType to find the IO function (so it does not really allow stacking errors).